chain/chain/src/store/merkle_proof.rs

```diff
@@ -0,0 +1,221 @@
+                    cur_index,
+                    level,
+                    counter,
+                    tree_size,
+                    &mut tree_nodes,
+                )?
+            } else {
+                self.get_merkle_tree_node(cur_index, level, counter, tree_size, &mut tree_nodes)?
+            };
+            if let Some(hash) = maybe_hash {
+                path.push(MerklePathItem { hash, direction });
+            }
+            cur_index /= 2;
+            iter = (iter + 1) / 2;
+            level += 1;
+            counter *= 2;
+        }
+        Ok(path)
+    }
+
+    /// Get node at given position (index, level). If the node does not exist, return `None`.
+    fn get_merkle_tree_node(
+        &self,
+        index: u64,
+        level: u64,
+        counter: u64,
+        tree_size: u64,
+        tree_nodes: &mut HashMap<(u64, u64), Option<MerkleHash>>,
+    ) -> Result<Option<MerkleHash>, Error> {
+        if let Some(hash) = tree_nodes.get(&(index, level)) {
+            Ok(*hash)
+        } else {
+            if level == 0 {
+                let maybe_hash = if index >= tree_size {
+                    None
+                } else {
+                    Some(self.get_block_hash_from_ordinal(index)?)
+                };
+                tree_nodes.insert((index, level), maybe_hash);
+                Ok(maybe_hash)
+            } else {
+                let cur_tree_size = (index + 1) * counter;
+                let maybe_hash = if cur_tree_size > tree_size {
+                    if index * counter <= tree_size {
+                        let left_hash = self.get_merkle_tree_node(
+                            index * 2,
+                            level - 1,
+                            counter / 2,
+                            tree_size,
+                            tree_nodes,
+                        )?;
+                        let right_hash = self.reconstruct_merkle_tree_node(
+                            index * 2 + 1,
+                            level - 1,
+                            counter / 2,
+                            tree_size,
+                            tree_nodes,
+                        )?;
+                        combine_maybe_hashes(left_hash, right_hash)
+                    } else {
+                        None
+                    }
+                } else {
+                    Some(
+                        *self
+                            .get_block_merkle_tree_from_ordinal(cur_tree_size)?
+                            .get_path()
+                            .last()
+                            .ok_or_else(|| Error::Other("Merkle tree node missing".to_string()))?,
+                    )
+                };
+                tree_nodes.insert((index, level), maybe_hash);
+                Ok(maybe_hash)
+            }
+        }
+    }
+
+    /// Reconstruct node at given position (index, level). If the node does not exist, return `None`.
+    fn reconstruct_merkle_tree_node(
```

Comment:
Consider moving this and some of the above functions outside of the Trait. Otherwise, these functions become public and pollute the namespace of what's available on Store. You can pass in the &impl trait in place of a &self.